src/NuGet.Core/NuGet.Build.Tasks/StaticGraphRestoreTaskBase.cs

```diff
@@ -0,0 +1,272 @@
+                using (var loggingQueue = new TaskLoggingQueue(Log))
+                using (var process = new Process())
+                {
+                    process.EnableRaisingEvents = true;
+                    process.StartInfo = new ProcessStartInfo
+                    {
+                        Arguments = GetCommandLineArguments(MSBuildBinPath),
+                        CreateNoWindow = true,
+                        FileName = GetProcessFileName(ProcessFileName),
+                        RedirectStandardInput = true,
+                        RedirectStandardOutput = true,
+                        UseShellExecute = false,
+                        WorkingDirectory = Environment.CurrentDirectory,
+                    };
+
+                    // Place the output in the queue which handles logging messages coming through on StdOut
+                    process.OutputDataReceived += (sender, args) => loggingQueue.Enqueue(args?.Data);
+
+                    process.Exited += (sender, args) => semaphore.Release();
+
+                    try
+                    {
+                        Log.LogMessage(MessageImportance.Low, "Running command: \"{0}\" {1}", process.StartInfo.FileName, process.StartInfo.Arguments);
+
+                        process.Start();
+
+                        WriteArguments(process.StandardInput);
+
+                        process.BeginOutputReadLine();
+
+                        semaphore.Wait(_cancellationTokenSource.Token);
+
+                        if (!process.HasExited)
+                        {
+                            try
+                            {
+                                process.Kill();
+                            }
+                            catch (InvalidOperationException)
+                            {
+                                // The process may have exited, in this case ignore the exception
+                            }
+                        }
+                    }
+                    catch (Exception e) when (
+                        e is OperationCanceledException
+                        || (e is AggregateException aggregateException && aggregateException.InnerException is OperationCanceledException))
+                    {
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Log.LogErrorFromException(e);
+            }
+
+            return !Log.HasLoggedErrors;
+        }
+
+        /// <summary>
+        /// Gets the command-line arguments to use when launching the process that executes the restore.
+        /// </summary>
+        internal string GetCommandLineArguments(string msbuildBinPath)
+        {
+            return string.Concat(
+#if IS_CORECLR
+                "\"", Path.Combine(ThisAssemblyLazy.Value.DirectoryName, Path.ChangeExtension(ThisAssemblyLazy.Value.Name, ".Console.dll")), "\"",
+                " ", "\"", Path.Combine(msbuildBinPath, "MSBuild.dll"), "\"",
```

Comment:
minor nitpick/suggestion

```suggestion
                " \"", Path.Combine(msbuildBinPath, "MSBuild.dll"), "\"",
```